File: Java/Treaps.java
import java.util.Random;

class Node {
    int key;
    int priority;
    Node left;
    Node right;
}

public class Treap {
    private Node createNode(int key) {
        Node newNode = new Node();
        newNode.key = key;
        newNode.priority = new Random().nextInt();
        newNode.left = newNode.right = null;
        return newNode;
    }

    private Node rotateRight(Node y) {
        Node x = y.left;
        Node T2 = x.right;

        x.right = y;
        y.left = T2;

        return x;
    }

    private Node rotateLeft(Node x) {
        Node y = x.right;
        Node T2 = y.left;

        y.left = x;
        x.right = T2;

        return y;
    }

    private Node insert(Node root, int key) {
        if (root == null) return createNode(key);

        if (key <= root.key) {
            root.left = insert(root.left, key);
            if (root.left.priority > root.priority)
                root = rotateRight(root);
        } else {
            root.right = insert(root.right, key);
            if (root.right.priority > root.priority)
                root = rotateLeft(root);
        }

        return root;
    }

    private void inorderTraversal(Node root) {
        if (root != null) {
            inorderTraversal(root.left);
            System.out.print("(" + root.key + ", " + root.priority + ") ");
            inorderTraversal(root.right);
        }
    }

    public static void main(String[] args) {
        Node root = null;

        Treap treap = new Treap();
        root = treap.insert(root, 50);
        root = treap.insert(root, 30);
        root = treap.insert(root, 20);
        root = treap.insert(root, 40);
        root = treap.insert(root, 70);
        root = treap.insert(root, 60);
        root = treap.insert(root, 80);

        System.out.println("Inorder traversal of the treap:");
        treap.inorderTraversal(root);
        System.out.println();
    }
}

File: Treaps.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

struct Node {
    int key;
    int priority;
    struct Node* left;
    struct Node* right;
};

struct Node* createNode(int key) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    if (!newNode) {
        perror("Memory allocation failed");
        exit(EXIT_FAILURE);
    }
    newNode->key = key;
    newNode->priority = rand();
    newNode->left = newNode->right = NULL;
    return newNode;
}

struct Node* rotateRight(struct Node* y) {
    struct Node* x = y->left;
    struct Node* T2 = x->right;

    x->right = y;
    y->left = T2;

    return x;
}

struct Node* rotateLeft(struct Node* x) {
    struct Node* y = x->right;
    struct Node* T2 = y->left;

    y->left = x;
    x->right = T2;

    return y;
}

struct Node* insert(struct Node* root, int key) {
    if (!root) return createNode(key);

    if (key <= root->key) {
        root->left = insert(root->left, key);
        if (root->left->priority > root->priority)
            root = rotateRight(root);
    } else {
        root->right = insert(root->right, key);
        if (root->right->priority > root->priority)
            root = rotateLeft(root);
    }

    return root;
}

void inorderTraversal(struct Node* root) {
    if (root) {
        inorderTraversal(root->left);
        printf("(%d, %d) ", root->key, root->priority);
        inorderTraversal(root->right);
    }
}

int main() {
    struct Node* root = NULL;
    srand(time(NULL));

    root = insert(root, 50);
    root = insert(root, 30);
    root = insert(root, 20);
    root = insert(root, 40);
    root = insert(root, 70);
    root = insert(root, 60);
    root = insert(root, 80);

    printf("Inorder traversal of the treap:\n");
    inorderTraversal(root);
    printf("\n");

    return 0;
}
